Acceleration Methods and Clean up

csdechant opened the pull request:
This PR adds acceleration methods and periodic variable difference calculations. The following is a list of the added objects:
- Kernels:
     - AccelerationByAveraging
     - ShootMethodLog
     - CoeffDiffusionForShootMethod
     - ReactionSecondOrderLogForShootMethod
     - ReactionThirdOrderLogForShootMethod
- AuxKernels:
     - DensityNormalization
- Postprocessors:
     - AverageNodalDensity
     - AverageNodalDifference
     - PeriodicComparisonCounter
     - RelativeElementL2Difference
- Actions
     - DriftDiffusionAction
     - PeriodicControllers
     - PeriodicRelativeNodalDifference

Also with this PR, the Zapdos Tutorial was updated (sections: 2.2.16, 2.2.17, 3.4, and 4.4.2) and old input files in the problem directory have been modified to run (though files mean_en_post_PJFNK, mean_en_postprocessor, and TM do not converge).

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/499910/) on 10575b6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/zapdos/docs/PRs/55/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/zapdos/docs/PRs/55/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d407b7496d0661c18e79be316f1671a8d9ba2489`

cticenhour commented:
@csdechant I'll add the comments I noted when I looked at your branch the other day once I get a chance, probably in an hour or so.

keniley1 commented on problems/RF_Plasma.i:
```diff
@@ -0,0 +1,357 @@
+    position_units = ${dom0Scale}
+    Additional_Outputs = 'ElectronTemperature Current EField'
+  [../]
+[]
```
Is it possible to write multiple blocks here for a single domain? e.g. [./plasma_nitrogen], [./plasma_oxygen], etc. for an air plasma, all in the same block). I also have a drift diffusion action that I've been using (yours looks better so I'll be replacing it when this PR goes through), but I've been working with input files that have ~30 species and I found that it's much easier for a user readability standpoint to be able to split up those variables by species instead of having "charged_particle" be a list of 30 species. It's easier to read that way.

Not a necessary thing, just curious.

csdechant replied:
Not currently, since the action adds the potential "diffusion" to each block it is used in. So calling it in multiple times for in one block will cause that block to have multiple potential diffusion kernels. I can add an bool statement like "add_potential_diffusion", which should fix that and then you can add the action multiple times in one block.

cticenhour commented on src/actions/AddPeriodicControllers.C:
```diff
@@ -0,0 +1,395 @@
+          params.set<std::vector<std::string>>("enable_objects") = _enable_start;
+          params.set<std::vector<Real>>("start_time") = _enable_start_start_time_index;
+          params.set<std::vector<Real>>("end_time") = _enable_start_end_time_index;
+          params.set<ExecFlagEnum>("execute_on", true) = {EXEC_INITIAL, EXEC_TIMESTEP_BEGIN};
```
Is there any way to optimize these if statements a bit through this section of the code? It seems as though you are setting the same parameters for the same values for `if true` and then `else` (like `execute_on` and `set_sync_times`). I would pull common set points out one level so that you're doing less repetitive work, and it makes readability better since you immediately see the difference in the cases.

cticenhour commented on src/actions/AddPeriodicRelativeNodalDifference.C:
```diff
@@ -0,0 +1,434 @@
+          params.set<std::vector<std::string>>("enable_objects") = {_enable_start};
+          params.set<std::vector<Real>>("start_time") = _enable_start_start_time_index;
+          params.set<std::vector<Real>>("end_time") = _enable_start_end_time_index;
+          params.set<ExecFlagEnum>("execute_on", true) = {EXEC_INITIAL, EXEC_TIMESTEP_BEGIN};
```
See my comment from `AddPeriodicControllers.C`. Same applies in the if statements in this file.

csdechant commented on doc/zapdos_content/syntax/Lymberopoulos_with_argon_metastables.i:
```diff
@@ -0,0 +1,871 @@
+dom0Scale=25.4e-3
```
The reasoning was to try to give the reader all the information in one spot (or one pdf file), but that reference code is old (uses old CRANE kernels). I agree, I would be better to just reference the test files.

csdechant commented on doc/Zapdos_Tutorial.tex:
```diff
@@ -0,0 +1,600 @@
+  \tableofcontents
+
+  \chapter{Introduction} \label{introduction}
+  Zapdos is an open source plasma fluid code developed by Dr. Alexander Lindsay at North Carolina State University (NCSU). It is built as an application in the MOOSE (Multiphysics Object Oriented Simulation Environment) framework. The MOOSE framework is an object-oriented C++ finite element method (FEM) framework developed at Idaho National Laboratory for the purpose of solving highly coupled mul ...
```
Will change. Also, it was Alex's suggestion during GEC to include the LaTex file with a make file, I just now adding it.

csdechant commented on problems/TM.i:
```diff
@@ -1,3 +1,5 @@
+#Runs but Solve failed
```
It converges for the first time step, then nonlinear solve diverges due to 'DIVERGED_LINE_SEARCH'. I am not sure why, but it the time I was just trying to make sure all the input files can at least run.

cticenhour replied:
That's fair....we need to look over more of these to make sure they work well if we're going to keep them around. I don't want it to hold this PR up. It was good to put a comment at the top of each describing the behavior - thanks!

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/501119/) on 2419c69 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/zapdos/docs/PRs/55/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/zapdos/docs/PRs/55/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d407b7496d0661c18e79be316f1671a8d9ba2489`

csdechant commented on problems/mean_en_post_PJFNK.i:
```diff
@@ -1,3 +1,5 @@
+#Runs but Solve failed
```
I was trying to do as you suggested and the input file is running, but the dt is staying around ~2e-10 (so the file is not done yet). The only difference between the two files are the potential BCs (mean_en has NeumannCircuitVoltageMoles_KV, while mean_en_PJFNK has CircuitDirichletPotential with the postprocessor) and their solve type (Newton and PJFNK). Maybe optimizing the PetsC options would help it run better, but I need to look into the options usually used for PJFNK.

cticenhour replied:
Okay, those differences make sense. So you do not get the "solve failed" that occurred in the previous iteration? If these small dt's were not expected behavior, then massaging the solve options may help. I would be wary of going down a rabbit hole here though. This could be work for a separate PR where we're cleaning up the problems directory.

csdechant commented:
I added the changes suggested by @cticenhour and added the option to add the drift-diffusion action more than once per block requested by @keniley1. The only thing that needs future work is to get some of the input file in the problem directory to converge (they all can run though).

cticenhour commented:
@keniley1  do these changes fit what you wanted? If so, I'm ready to approve/merge. The issue files in the problems directory should go in a future PR (in my opinion).

keniley1 commented:
Yeah, I'm happy with the changes! I agree that the problems directory is a separate issue altogether. No need to cram more in to this PR.
